docroot/profiles/custom/cgov_site/modules/custom/pdq_core/src/Plugin/rest/resource/PDQResource.php

```diff
@@ -0,0 +1,229 @@
+   * Groups of 25 at a time should be safe.
+   *
+   * @param array $summaries
+   *   Sequence of node ID + language code pairs.
+   *
+   * @return \Drupal\rest\ModifiedResourceResponse
+   *   Possibly empty sequence of node ID/language/error arrays
+   *
+   * @throws \Symfony\Component\HttpKernel\Exception\HttpException
+   *   Throws exception expected.
+   */
+  public function post(array $summaries) {
+
+    $errors = [];
+    foreach ($summaries as $summary) {
+      try {
+        list($nid, $language) = $summary;
+        $node = Node::load($nid);
+        $entity = $node->getTranslation($language);
+        $entity->moderation_state = 'published';
+        $entity->save();
+        unset($entity);
+        unset($node);
+      }
+      catch (Exception $e) {
+        $message = $e->getMessage() ?? 'unexpected failure';
+        $errors[] = [$nid, $language, $message];
+      }
+    }
+    return new ModifiedResourceResponse(['errors' => $errors], 200);
+  }
+
+  /**
+   * Responds to DELETE requests.
+   *
+   * @param string $id
+   *   The CDR ID of a PDQ Summary document (optionally prefixed by
+   *   "CDR").
+   *
+   * @return \Drupal\rest\ModifiedResourceResponse
+   *   The HTTP response object.
+   *
+   * @throws \Symfony\Component\HttpKernel\Exception\HttpException
+   *   Throws exception expected.
+   */
+  public function delete($id) {
+
+    // @todo: avoid deleting target of links from other content.
+    // Make sure the client gave us something to look for.
+    if (!$id) {
+      throw new BadRequestHttpException(t('No ID was provided'));
+    }
+
+    // Make sure the CDR ID matches exactly one entity.
+    $matches = $this->lookupCdrId($id);
+    if (count($matches) === 0) {
+      $msg = t('CDR ID @id not found', ['@id' => $id]);
+      throw new NotFoundHttpException($msg);
+    }
+    if (count($matches) > 1) {
+      $msg = t('Ambiguous CDR ID @id', ['@id' => $id]);
+      throw new BadRequestHttpException($msg);
+    }
+    list($nid, $langcode) = $matches[0];
+    $node = Node::load($nid);
+
+    // Apply deletion logic based on language.
+    if ($langcode === 'es') {
+      $node->removeTranslation('es');
+      $node->save();
+    }
+    else {
+      if ($node->hasTranslation('es')) {
+        throw new BadRequestHttpException(t('Spanish translation exists'));
+      }
+      $node->delete();
+    }
+    return new ModifiedResourceResponse(NULL, 204);
+  }
+
+  /**
+   * Find the Drupal entities which store a given PDQ document.
+   *
+   * @param string $id
+   *   The CDR ID of a PDQ Summary document (optionally prefixed by
+   *   "CDR").
+   *
+   * @return array
+   *   Sequence of value pairs, each of which has a node ID and a
+   *   language code.
+   */
+  private function lookupCdrId($id) {
+    if (substr($id, 0, 3) === 'CDR') {
+      $id = substr($id, 3);
+    }
+    $matches = [];
+    $query = \Drupal::database()->select('node__field_pdq_cdr_id', 'c');
```

Comment:
Possibly. I wanted a query which didn't depend on knowing anything about the specific bundles for the PDQ content types.

Reply:
Digging further (gently prodded by Brendan) I realize I was mistaken about how `EntityQuery` works. Here's the Slack exchange:

**Brendan**: curious about your pr. especially about why you directly queried the database instead of usinng a service like entityquery. Figure you must have some reason, thought it would be useful to understand for the future.

**Bob**: Excellent (implied) question. I had a bunch of reasons, not all of them very good.

1. More efficient
2. I have found over the years that it's easier to reverse-engineer the Drupal tables than to find comprehensive, useful Drupal documentation
3. I find Drupal's deprecation churn cycle even more annoying than Microsoft's ("learn and use this API -- no, wait, that's no good, use these instead..."), which in combination with the lack of a comprehensive, authoritative, easily identifiable set of documentation is perhaps Drupal's biggest flaw
4. I mistakenly thought that the query APIs depended on knowing the content type ("bundle") and not just the entity type ("node" in this case).

Now that I realize that #4 (which I had thought of as the only "real" reason) is wrong, I'll rewrite to use the service.


Thanks to @bryanpizzillo and @BrendanBeltzNIH for catching this and nudging me to revisit it. 👍 Always learning! 😃

Reply:
DB query replaced with entity query.